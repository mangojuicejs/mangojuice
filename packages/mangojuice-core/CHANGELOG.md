# Change Log

All notable changes to this project will be documented in this file.
See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.1.3"></a>
## 0.1.3 (2017-07-12)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))
* **core:** recursive nesting works now ([c462be5](https://github.com/mangojuicejs/mangojuice/commit/c462be5))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.3"></a>
## 0.1.3 (2017-07-08)



<a name="0.1.3"></a>
## 0.1.3 (2017-07-08)



<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.3"></a>
## 0.1.3 (2017-07-08)



<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)



<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)
