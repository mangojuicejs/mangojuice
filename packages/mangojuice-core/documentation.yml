toc:
  - name: Basic
    description: |
      The most common APIs that you will work with most of the time
    children:
      - Block
      - LogicBase
      - cmd
      - run
      - mount
      - task
      - cancel
      - child
      - depends
  - name: Extras
    description: |
      Other pieces of MangoJuice
    children:
      - Process
      - Command
      - TaskMeta
      - DefaultLogger
      - ComputedField
      - Mounter
      - defineCommand
      - ensureCommand
      - decorateLogic
      - handleBefore
      - handleAfter
      - handle
      - call
      - logicOf
      - observe
      - procOf
      - bind
      - hydrate
      - delay
