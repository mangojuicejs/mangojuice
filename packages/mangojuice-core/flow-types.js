// @flow
type CommandType<M> = {
  is: (...args: any) => bool,
  model: M,
};

type TaskObject = {
  success: (cmd: any) => TaskObject,
  fail: (cmd: any) => TaskObject,
  args: (...args: Array<any>) => TaskObject,
  engine: (exec: any) => TaskObject
};

type DependsType = {
  compute: (Function) => DependsType
};

type LogicConfig<MetaType> = {
  initCommands?: any,
  meta?: MetaType
};

type ConfiguredLogic = {
};

type BlockType = {
  createModel: (...args: any) => Object,
  Logic: Class<*>,
  View?: any
};

declare module "mangojuice-core" {
  declare export function cmd(): any;

  declare export function debounce(): any;

  declare export function throttle(): any;

  declare export function task(fn: Function): TaskObject;

  declare export function delay(ms: number): Promise<void>;

  declare export function cancel(cmd: any): any;

  declare export function logicOf<T>(model: Object, clazz?: Class<T>): T;

  declare export function child<T>(logicClass: Class<{ +config?: (...args: T) => any }>, ...args: T): ConfiguredLogic;

  declare export function depends(...args: any): DependsType;

  declare export function observe(model: Object, destroyeded: Promise<any>, handler: Function): void;

  declare export function run(block: BlockType, opts?: Object): any;

  declare export function bind(block: BlockType, opts?: Object): any;

  declare export function hydrate(block: BlockType, model: Object): any;

  declare export var utils: any;

  declare export class LogicBase<ModelType = any, SharedType = any, MetaType = any> {
    +model: ModelType;
    +shared: SharedType;
    +meta: MetaType;

    hub(cmd: CommandType<*>): ?Promise<any>;

    hubBefore(cmd: CommandType<*>): ?Promise<any>;

    hubAfter(cmd: CommandType<*>): ?Promise<any>;

    port(exec: (cmd: any) => Promise<any>, destroyed: Promise<any>): ?Promise<any>;

    children(): {[k: $Enum<ModelType>]: Class<*> | ConfiguredLogic};

    computed(): {[k: $Enum<ModelType>]: (() => any) | DependsType};

    config(...args: any): LogicConfig<MetaType>;
  }

  declare export default {|
    +LogicBase: typeof LogicBase,
    +cmd: typeof cmd,
    +debounce: typeof debounce,
    +throttle: typeof throttle,
    +task: typeof task,
    +delay: typeof delay,
    +cancel: typeof cancel,
    +logicOf: typeof logicOf,
    +child: typeof child,
    +depends: typeof depends,
    +observe: typeof observe,
    +run: typeof run,
    +bind: typeof bind,
    +hydrate: typeof hydrate,
    +utils: typeof utils
  |};
}
