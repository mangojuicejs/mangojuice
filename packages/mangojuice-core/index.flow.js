// @flow

type NestType<A, T, Y, U> = {
  +logic: A,
  args: (c1?: T, c2?: Y, c3?: U) => NestType<A, T, Y, U>,
  handler: () => NestType<A, T, Y, U>,
  singleton: (?Boolean) => NestType<A, T, Y, U>
};

type DependsType = {
  compute: (Function) => DependsType
};

interface ConfigFunction<MetaType, T, Y, U> {
  +config?: (c1?: T, c2?: Y, c3?: U) => ConfigType<MetaType>;
};

type ConfigType<M> = {
  initCommands?: any,
  meta?: M
};

declare class LogicBase<ModelType = any, SharedType = any, MetaType = any> {
  +model: ModelType;
  +shared: SharedType;
  +meta: MetaType;
  +destroy: Promise<any>;

  port(): ?Promise<any>;

  children(): {[key: $Enum<ModelType>]: NestType<*, *, *, *>};

  computed(): {[k: $Enum<ModelType>]: () => any | DependsType};

  exec(): Promise<any>;

  depends(...deps: Array<Object>): DependsType;

  nest<T, Y, U, L: ConfigFunction<MetaType, T, Y, U>&LogicBase<*,SharedType,*>>(logic: L): NestType<L, T, Y, U>;
}
