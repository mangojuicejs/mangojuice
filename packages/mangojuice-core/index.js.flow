// @flow

// Cmd
//
declare export var Cmd: {
  update: any,
  batch: any,
  execLatest: any,
  execEvery: any,
  debounce: any,
  throttle: any
};

// Task
//
type TaskObject = {
  success: (cmd: any) => TaskObject,
  fail: (cmd: any) => TaskObject
};
declare export var Task: {
  create: (Function) => TaskObject;
};


// Run
//
declare export var Run: any;


// LogicBase
//
type NestType<A, T, Y, U> = {
  +logic: A,
  args: (c1?: T, c2?: Y, c3?: U) => NestType<A, T, Y, U>,
  handler: () => NestType<A, T, Y, U>,
  singleton: (?Boolean) => NestType<A, T, Y, U>
};

type DependsType = {
  compute: (Function) => DependsType
};

interface ConfigFunction<MetaType, T, Y, U> {
  +config?: (c1?: T, c2?: Y, c3?: U) => ConfigType<MetaType>;
};

type ConfigType<M> = {
  initCommands?: any,
  meta?: M
};

declare export class LogicBase<ModelType = any, SharedType = any, MetaType = any> {
  +model: ModelType;
  +shared: SharedType;
  +meta: MetaType;
  +destroy: Promise<any>;

  port(): ?Promise<any>;

  children(): {[key: $Enum<ModelType>]: NestType<*, *, *, *>};

  computed(): {[k: $Enum<ModelType>]: () => any | DependsType};

  exec(): Promise<any>;

  depends(...deps: Array<Object>): DependsType;

  nest<T, Y, U, L: ConfigFunction<MetaType, T, Y, U>&LogicBase<*,SharedType,*>>(logic: L): NestType<L, T, Y, U>;
}
