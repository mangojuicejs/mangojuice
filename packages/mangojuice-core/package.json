{
  "name": "mangojuice-core",
  "version": "0.1.6",
  "description": "",
  "author": {
    "name": "Artem Artemev",
    "email": "viiv.c58@gmail.com"
  },
  "bugs": {
    "url": "https://github.com/mangojuicejs/mangojuice/issues"
  },
  "homepage": "https://github.com/mangojuicejs/mangojuice#readme",
  "repository": "https://github.com/mangojuicejs/mangojuice",
  "license": "MIT",
  "main": "index.js",
  "rollup": {
    "moduleName": "Mjs"
  }
}
