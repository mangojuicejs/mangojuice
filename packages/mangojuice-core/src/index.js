import Process from './classes/Process';
import Command from './classes/Command';
import LogicBase from './classes/LogicBase';
import DefaultLogger from './classes/DefaultLogger';
import cmd from './core/cmd/cmd';
import cancel from './core/cmd/cancel';
import defineCommand from './core/cmd/defineCommand';
import ensureCommand from './core/cmd/ensureCommand';
import child from './core/logic/child';
import depends from './core/logic/depends';
import observe from './core/logic/observe';
import { handle, handleAfter, handleBefore } from './core/logic/handle';
import logicOf from './core/logic/logicOf';
import procOf from './core/logic/procOf';
import decorateLogic from './core/logic/decorateLogic';
import delay from './core/task/delay';
import task from './core/task/task';
import callTask from './core/task/callTask';
import { CANCEL } from './classes/AsyncTask';
import run from './core/run/run';
import bind from './core/run/bind';
import mount from './core/run/mount';
import hydrate from './core/run/hydrate';
import * as utils from './core/utils';
import * as config from './config';


export {
  // Constants
  CANCEL,
  // Classes
  Process,
  Command,
  LogicBase,
  DefaultLogger,
  // Core
  cmd,
  cancel,
  defineCommand,
  ensureCommand,
  child,
  depends,
  observe,
  handle,
  handleAfter,
  handleBefore,
  logicOf,
  procOf,
  decorateLogic,
  delay,
  task,
  callTask,
  run,
  bind,
  mount,
  hydrate,
  utils,
  config
};
