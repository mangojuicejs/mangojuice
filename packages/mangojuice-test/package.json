{
  "name": "mangojuice-test",
  "version": "2.0.0-y.0",
  "description": "",
  "author": {
    "name": "Artem Artemev",
    "email": "viiv.c58@gmail.com"
  },
  "bugs": {
    "url": "https://github.com/mangojuicejs/mangojuice/issues"
  },
  "homepage": "https://github.com/mangojuicejs/mangojuice#readme",
  "repository": "https://github.com/mangojuicejs/mangojuice",
  "license": "MIT",
  "main": "index.js",
  "rollup": {
    "moduleName": "Mjs.Test",
    "bundledDependencies": [],
    "moduleGlobals": {
      "mangojuice": "Mjs"
    }
  },
  "dependencies": {
    "mangojuice-core": "2.0.0-y.0"
  }
}
