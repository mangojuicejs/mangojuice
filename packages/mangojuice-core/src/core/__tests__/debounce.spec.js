import { child, context, logicOf, procOf, debounce, utils } from 'mangojuice-core';
import { runWithTracking } from 'mangojuice-test';


describe('debounce', () => {
  class TestLogic {
    create() {
      return { a: 0 };
    }
    cancelInc() {
      return debounce(100, this.increment).cancel();
    }
    incrementLeading(inc) {
      return debounce(100, this.increment, { leading: true, trailing: false }).args(inc);
    }
    incrementTrailing(inc) {
      return debounce(100, this.increment, { leading: false, trailing: true }).args(inc);
    }
    incrementLeadingTrailing(inc) {
      return debounce(100, this.increment, { leading: true, trailing: true }).args(inc);
    }
    increment(inc) {
      return { a: this.model.a + inc };
    }
  }

  describe('#create', () => {
    it('should invoke only trailing', async () => {
      const { app, commandNames } = await runWithTracking({ app: { Logic: TestLogic } });

      const res = app.proc.exec(logicOf(app.model).incrementTrailing(10));
      await utils.delay(10);
      expect(commandNames).toMatchSnapshot();
      expect(app.model).toEqual({ a: 0 });

      await utils.delay(100);
      expect(app.model).toEqual({ a: 10 });
      expect(commandNames).toMatchSnapshot();
      await res;
    });

    it('should delay trailing invoke after multiple calls', async () => {
      const { app, commandNames } = await runWithTracking({ app: { Logic: TestLogic } });

      app.proc.exec(logicOf(app.model).incrementTrailing(10));
      await utils.delay(80);
      app.proc.exec(logicOf(app.model).incrementTrailing(43));
      await utils.delay(80);
      expect(commandNames).toMatchSnapshot();
      expect(app.model).toEqual({ a: 0 });

      await utils.delay(100);
      expect(app.model).toEqual({ a: 43 });
      expect(commandNames).toMatchSnapshot();
    });

    it('should resolve promise after trailing invokation', async () => {
      const { app, commandNames } = await runWithTracking({ app: { Logic: TestLogic } });

      app.proc.exec(logicOf(app.model).incrementTrailing(10));
      await utils.delay(80);
      const res = app.proc.exec(logicOf(app.model).incrementTrailing(43));
      await utils.delay(50);
      expect(commandNames).toMatchSnapshot();
      expect(app.model).toEqual({ a: 0 });

      await res;
      expect(app.model).toEqual({ a: 43 });
      expect(commandNames).toMatchSnapshot();
    });

    it('should have only leading invokation', async () => {
      const { app, commandNames } = await runWithTracking({ app: { Logic: TestLogic } });

      app.proc.exec(logicOf(app.model).incrementLeading(10));
      await utils.delay(80);
      const res = app.proc.exec(logicOf(app.model).incrementLeading(43));
      await utils.delay(50);
      expect(commandNames).toMatchSnapshot();
      expect(app.model).toEqual({ a: 10 });

      await res;
      expect(app.model).toEqual({ a: 10 });
      expect(commandNames).toMatchSnapshot();
    });

    it('should have leading invokation and no trailing if invoked once', async () => {
      const { app, commandNames } = await runWithTracking({ app: { Logic: TestLogic } });

      const res = app.proc.exec(logicOf(app.model).incrementLeadingTrailing(10));
      expect(commandNames).toMatchSnapshot();
      expect(app.model).toEqual({ a: 10 });

      await res;
      expect(app.model).toEqual({ a: 10 });
    });

    it('should leading and trailing invokations', async () => {
      const { app, commandNames } = await runWithTracking({ app: { Logic: TestLogic } });

      app.proc.exec(logicOf(app.model).incrementLeadingTrailing(10));
      expect(commandNames).toMatchSnapshot();
      expect(app.model).toEqual({ a: 10 });

      app.proc.exec(logicOf(app.model).incrementLeadingTrailing(20));
      const res = app.proc.exec(logicOf(app.model).incrementLeadingTrailing(30));
      expect(app.model).toEqual({ a: 10 });

      await res;
      expect(app.model).toEqual({ a: 40 });
    });
  });

  describe('#cancel', () => {
    it('should cancel trailing invokation', async () => {
      const { app, commandNames } = await runWithTracking({ app: { Logic: TestLogic } });

      app.proc.exec(logicOf(app.model).incrementLeadingTrailing(10));
      expect(commandNames).toMatchSnapshot();
      expect(app.model).toEqual({ a: 10 });

      app.proc.exec(logicOf(app.model).incrementLeadingTrailing(20));
      app.proc.exec(logicOf(app.model).incrementLeadingTrailing(30));
      app.proc.exec(logicOf(app.model).cancelInc);
      expect(app.model).toEqual({ a: 10 });

      await app.proc.finished();
      expect(app.model).toEqual({ a: 10 });
    });
  });
});
