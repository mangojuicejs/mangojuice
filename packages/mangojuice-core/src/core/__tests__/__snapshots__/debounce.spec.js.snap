// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debounce #cancel should cancel trailing invokation 1`] = `
Array [
  Object {
    "a": 0,
  },
  TaskCmd {
    "customArgs": Array [
      10,
    ],
    "customExecId": "DELAYED",
    "execEvery": true,
    "executor": [Function],
    "id": 3,
    "metaObj": Object {
      "options": Object {
        "leading": true,
        "trailing": true,
      },
      "wait": 100,
    },
    "notifyCmd": [Function],
    "task": [Function],
  },
  Object {
    "a": 10,
  },
]
`;

exports[`debounce #create should delay trailing invoke after multiple calls 1`] = `
Array [
  Object {
    "a": 0,
  },
  TaskCmd {
    "customArgs": Array [
      10,
    ],
    "customExecId": "DELAYED",
    "execEvery": true,
    "executor": [Function],
    "id": 3,
    "metaObj": Object {
      "options": Object {
        "leading": false,
        "trailing": true,
      },
      "wait": 100,
    },
    "notifyCmd": [Function],
    "task": [Function],
  },
  TaskCmd {
    "customArgs": Array [
      43,
    ],
    "customExecId": "DELAYED",
    "execEvery": true,
    "executor": [Function],
    "id": 3,
    "metaObj": Object {
      "options": Object {
        "leading": false,
        "trailing": true,
      },
      "wait": 100,
    },
    "notifyCmd": [Function],
    "task": [Function],
  },
]
`;

exports[`debounce #create should delay trailing invoke after multiple calls 2`] = `
Array [
  Object {
    "a": 0,
  },
  TaskCmd {
    "customArgs": Array [
      10,
    ],
    "customExecId": "DELAYED",
    "execEvery": true,
    "executor": [Function],
    "id": 3,
    "metaObj": Object {
      "options": Object {
        "leading": false,
        "trailing": true,
      },
      "wait": 100,
    },
    "notifyCmd": [Function],
    "task": [Function],
  },
  TaskCmd {
    "customArgs": Array [
      43,
    ],
    "customExecId": "DELAYED",
    "execEvery": true,
    "executor": [Function],
    "id": 3,
    "metaObj": Object {
      "options": Object {
        "leading": false,
        "trailing": true,
      },
      "wait": 100,
    },
    "notifyCmd": [Function],
    "task": [Function],
  },
  Object {
    "a": 43,
  },
]
`;

exports[`debounce #create should have leading invokation and no trailing if invoked once 1`] = `
Array [
  Object {
    "a": 0,
  },
  TaskCmd {
    "customArgs": Array [
      10,
    ],
    "customExecId": "DELAYED",
    "execEvery": true,
    "executor": [Function],
    "id": 3,
    "metaObj": Object {
      "options": Object {
        "leading": true,
        "trailing": true,
      },
      "wait": 100,
    },
    "notifyCmd": [Function],
    "task": [Function],
  },
  Object {
    "a": 10,
  },
]
`;

exports[`debounce #create should have only leading invokation 1`] = `
Array [
  Object {
    "a": 0,
  },
  TaskCmd {
    "customArgs": Array [
      10,
    ],
    "customExecId": "DELAYED",
    "execEvery": true,
    "executor": [Function],
    "id": 3,
    "metaObj": Object {
      "options": Object {
        "leading": true,
        "trailing": false,
      },
      "wait": 100,
    },
    "notifyCmd": [Function],
    "task": [Function],
  },
  Object {
    "a": 10,
  },
  TaskCmd {
    "customArgs": Array [
      43,
    ],
    "customExecId": "DELAYED",
    "execEvery": true,
    "executor": [Function],
    "id": 3,
    "metaObj": Object {
      "options": Object {
        "leading": true,
        "trailing": false,
      },
      "wait": 100,
    },
    "notifyCmd": [Function],
    "task": [Function],
  },
]
`;

exports[`debounce #create should have only leading invokation 2`] = `
Array [
  Object {
    "a": 0,
  },
  TaskCmd {
    "customArgs": Array [
      10,
    ],
    "customExecId": "DELAYED",
    "execEvery": true,
    "executor": [Function],
    "id": 3,
    "metaObj": Object {
      "options": Object {
        "leading": true,
        "trailing": false,
      },
      "wait": 100,
    },
    "notifyCmd": [Function],
    "task": [Function],
  },
  Object {
    "a": 10,
  },
  TaskCmd {
    "customArgs": Array [
      43,
    ],
    "customExecId": "DELAYED",
    "execEvery": true,
    "executor": [Function],
    "id": 3,
    "metaObj": Object {
      "options": Object {
        "leading": true,
        "trailing": false,
      },
      "wait": 100,
    },
    "notifyCmd": [Function],
    "task": [Function],
  },
]
`;

exports[`debounce #create should invoke only trailing 1`] = `
Array [
  Object {
    "a": 0,
  },
  TaskCmd {
    "customArgs": Array [
      10,
    ],
    "customExecId": "DELAYED",
    "execEvery": true,
    "executor": [Function],
    "id": 3,
    "metaObj": Object {
      "options": Object {
        "leading": false,
        "trailing": true,
      },
      "wait": 100,
    },
    "notifyCmd": [Function],
    "task": [Function],
  },
]
`;

exports[`debounce #create should invoke only trailing 2`] = `
Array [
  Object {
    "a": 0,
  },
  TaskCmd {
    "customArgs": Array [
      10,
    ],
    "customExecId": "DELAYED",
    "execEvery": true,
    "executor": [Function],
    "id": 3,
    "metaObj": Object {
      "options": Object {
        "leading": false,
        "trailing": true,
      },
      "wait": 100,
    },
    "notifyCmd": [Function],
    "task": [Function],
  },
  Object {
    "a": 10,
  },
]
`;

exports[`debounce #create should leading and trailing invokations 1`] = `
Array [
  Object {
    "a": 0,
  },
  TaskCmd {
    "customArgs": Array [
      10,
    ],
    "customExecId": "DELAYED",
    "execEvery": true,
    "executor": [Function],
    "id": 3,
    "metaObj": Object {
      "options": Object {
        "leading": true,
        "trailing": true,
      },
      "wait": 100,
    },
    "notifyCmd": [Function],
    "task": [Function],
  },
  Object {
    "a": 10,
  },
]
`;

exports[`debounce #create should resolve promise after trailing invokation 1`] = `
Array [
  Object {
    "a": 0,
  },
  TaskCmd {
    "customArgs": Array [
      10,
    ],
    "customExecId": "DELAYED",
    "execEvery": true,
    "executor": [Function],
    "id": 3,
    "metaObj": Object {
      "options": Object {
        "leading": false,
        "trailing": true,
      },
      "wait": 100,
    },
    "notifyCmd": [Function],
    "task": [Function],
  },
  TaskCmd {
    "customArgs": Array [
      43,
    ],
    "customExecId": "DELAYED",
    "execEvery": true,
    "executor": [Function],
    "id": 3,
    "metaObj": Object {
      "options": Object {
        "leading": false,
        "trailing": true,
      },
      "wait": 100,
    },
    "notifyCmd": [Function],
    "task": [Function],
  },
]
`;

exports[`debounce #create should resolve promise after trailing invokation 2`] = `
Array [
  Object {
    "a": 0,
  },
  TaskCmd {
    "customArgs": Array [
      10,
    ],
    "customExecId": "DELAYED",
    "execEvery": true,
    "executor": [Function],
    "id": 3,
    "metaObj": Object {
      "options": Object {
        "leading": false,
        "trailing": true,
      },
      "wait": 100,
    },
    "notifyCmd": [Function],
    "task": [Function],
  },
  TaskCmd {
    "customArgs": Array [
      43,
    ],
    "customExecId": "DELAYED",
    "execEvery": true,
    "executor": [Function],
    "id": 3,
    "metaObj": Object {
      "options": Object {
        "leading": false,
        "trailing": true,
      },
      "wait": 100,
    },
    "notifyCmd": [Function],
    "task": [Function],
  },
  Object {
    "a": 43,
  },
]
`;
